Module not found: Package path ./rsc is not exported from package

### Description
From examples/next-ai-rsc

When downloading the code from Github repo and running pnpm install + pnpm dev. And run into an module error:
Failed to compile
./components/llm-stocks/stock-purchase.tsx:4:0
Module not found: Package path ./rsc is not exported from package C:\Users\nicoz\Downloads\ai-main\ai-main\node_modules\.pnpm\node_modules\ai (see exports field in C:\Users\nicoz\Downloads\ai-main\ai-main\node_modules\.pnpm\node_modules\ai\package.json)
  2 |
  3 | import { useId, useState } from 'react';
> 4 | import { useActions, useAIState, useUIState } from 'ai/rsc';
  5 | import { formatNumber } from '@/lib/utils';
  6 |
  7 | import type { AI } from '../../app/action';

https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./components/llm-stocks/index.tsx 

![image](https://github.com/vercel/ai/assets/31444892/129e4c91-f95b-4e31-a77b-5af5939d851e)


### Code example

_No response_

### Additional context

_No response_

Hitting this issue on linux, yarn, node 21.6.2. Here's the build failure I encounter:
```
Type error: Cannot find module 'ai/rsc' or its corresponding type declarations.

> 1 | import { createAI } from 'ai/rsc';
```

If I run the following from my project root:
```bash
cp node_modules/ai/rsc/dist/rsc-types.d.ts node_modules/ai/rsc/dist/index.d.ts
```

Then the build completes fine. I see that we're specifying `rsc-types.d.ts` as the `types` value in `ai/package.json`'s exports block for `./rsc`, but perhaps some build tooling here doesn't respect this and is looking for `index.d.ts`?

--- a/.changeset/smooth-numbers-whisper.md
+++ b/.changeset/smooth-numbers-whisper.md
@@ -1,0 +1,5 @@
+---
+'ai': patch
+---
+
+fix type resolution

--- a/packages/core/package.json
+++ b/packages/core/package.json
@@ -40,44 +40,37 @@
     ".": {
       "types": "./dist/index.d.ts",
       "import": "./dist/index.mjs",
-      "module": "./dist/index.mjs",
       "require": "./dist/index.js"
     },
     "./rsc": {
-      "types": "./rsc/dist/index.d.mts",
+      "types": "./rsc/dist/index.d.ts",
       "react-server": "./rsc/dist/rsc-server.mjs",
-      "import": "./rsc/dist/rsc-client.mjs",
-      "module": "./rsc/dist/rsc-client.mjs"
+      "import": "./rsc/dist/rsc-client.mjs"
     },
     "./prompts": {
       "types": "./prompts/dist/index.d.ts",
       "import": "./prompts/dist/index.mjs",
-      "module": "./prompts/dist/index.mjs",
       "require": "./prompts/dist/index.js"
     },
     "./react": {
       "types": "./react/dist/index.d.ts",
       "react-server": "./react/dist/index.server.mjs",
       "import": "./react/dist/index.mjs",
-      "module": "./react/dist/index.mjs",
       "require": "./react/dist/index.js"
     },
     "./svelte": {
       "types": "./svelte/dist/index.d.ts",
       "import": "./svelte/dist/index.mjs",
-      "module": "./svelte/dist/index.mjs",
       "require": "./svelte/dist/index.js"
     },
     "./vue": {
       "types": "./vue/dist/index.d.ts",
       "import": "./vue/dist/index.mjs",
-      "module": "./vue/dist/index.mjs",
       "require": "./vue/dist/index.js"
     },
     "./solid": {
       "types": "./solid/dist/index.d.ts",
       "import": "./solid/dist/index.mjs",
-      "module": "./solid/dist/index.mjs",
       "require": "./solid/dist/index.js"
     }
   },

--- a/packages/core/rsc/package.json
+++ b/packages/core/rsc/package.json
@@ -1,5 +1,5 @@
 {
-  "types": "./dist/rsc-types.d.ts",
+  "types": "./dist/index.d.ts",
   "exports": {
     "types": "./dist/index.d.ts",
     "react-server": "./dist/rsc-server.mjs",

--- a/packages/core/rsc/package.json
+++ b/packages/core/rsc/package.json
@@ -1,10 +1,9 @@
 {
-  "types": "./dist/rsc-types.d.mts",
+  "types": "./dist/rsc-types.d.ts",
   "exports": {
-    "types": "./dist/index.d.mts",
+    "types": "./dist/index.d.ts",
     "react-server": "./dist/rsc-server.mjs",
-    "import": "./dist/rsc-client.mjs",
-    "module": "./dist/rsc-client.mjs"
+    "import": "./dist/rsc-client.mjs"
   },
   "private": true,
   "peerDependencies": {

--- a/packages/core/tsup.config.ts
+++ b/packages/core/tsup.config.ts
@@ -80,13 +80,27 @@ export default defineConfig([
     dts: true,
     sourcemap: true,
   },
-  // RSC APIs - server, client, types
+  // RSC APIs - server, client
   {
-    entry: ['rsc/rsc-server.ts', 'rsc/rsc-client.ts', 'rsc/index.ts'],
+    entry: ['rsc/rsc-server.ts', 'rsc/rsc-client.ts'],
     outDir: 'rsc/dist',
     format: ['esm'],
     external: ['react', 'zod', /\/rsc-shared/],
     dts: true,
     sourcemap: true,
+  },
+  // RSC APIs - types
+  {
+    entry: ['rsc/index.ts'],
+    outDir: 'rsc/dist',
+    dts: true,
+    outExtension() {
+      return {
+        // It must be `.d.ts` instead of `.d.mts` to support node resolution.
+        // See https://github.com/vercel/ai/issues/1028.
+        dts: '.d.ts',
+        js: '.mjs',
+      };
+    },
   },
 ]);

